velox/row/CompactRow.h

```diff
@@ -36,6 +36,16 @@ class CompactRow {
   /// 'buffer' must have sufficient capacity and set to all zeros.
   int32_t serialize(vector_size_t index, char* buffer);
 
+  /// Serializes rows in range [offset, offset + size) into 'buffer' at given
+  /// 'bufferOffsets'. 'buffer' must have sufficient capacity and set to all
+  /// zeros. The size of 'bufferOffsets' and 'size' must be the same.
+  /// 'bufferOffsets' will be updated by the actual bytes written.
+  void serialize(
+      vector_size_t offset,
+      vector_size_t size,
+      char* buffer,
+      std::vector<size_t>& bufferOffsets);
```

Comment:
I wonder how does this work at a high level? How do we know where to start writing 2nd, 3nd, n-th row? Do we pre-compute the offsets or do we expect the caller to pre-compute these? Would you clarify?

Reply:
@mbasmanova Updated the comment https://github.com/facebookincubator/velox/pull/10714/commits/9a8ad67847c4a23e79d0f0820cebba0aea62713f

"bufferOffsets" should be pre-computed by the caller. In Gluten, we would set the bufferOffsets for each row to `sizeof(rowSize) + rowSize` so that we can prepend the value of rowSize during serialization. This allows the deserialization process to know how many bytes to read for the next row.

Reply:
@marin-ma Thank you for clarifying and updating the comment.

> /// 'bufferOffsets' will be updated by the actual bytes written for each row.

Why do we need this? We expect this value to be the same as in the input, no?

Reply:
@mbasmanova Serialization is performed column by column. After serializing all rows for the first column, 'bufferOffsets' is updated with the bytes written. Then, the updated 'bufferOffsets' is used to serialize the next column. Therefore, the added value of 'bufferOffsets' should equal 'rowSize' for each row when all columns are serialized.

Reply:
Got it. I assume the caller of this API is not interested in these updates though. Hence, from API design perspective,  would it make sense to have bufferOffsets as const & and internally we can use a different 'offsets' vector.